src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3808 @@
+        if !(B in modules)
+          parent_hom[B] = H
+          push!(modules,B)
+        end
+      end
+      if B===N
+        found_N = true
+        break
+      end
+    end
+    if found_N
+      break
+    end
+  end
+  if !found_N
+    throw(DomainError("There is no path of canonical homomorphisms between the modules!"))
+  end
+  result = v
+  A = N
+  while A !== M
+    H = parent_hom[A]
+    result = H(result)
+    A = codomain(H)
+  end
+  return result
+end
+
+function all_canonical_maps(M::SubQuo, N::SubQuo)
+  # from N to M
+
+  all_paths = []
+
+  function helper_dfs(U::SubQuo, D::SubQuo, visited::Set, path::Vector)
+    if U === D
+      push!(all_paths, path)
+      return
+    end
+    for neighbor_morphism in U.outgoing_morphisms
+      if !(neighbor_morphism in visited)
+
+        #push!(visited, neighbor_morphism)
+        #push!(path, neighbor_morphism)
+        #helper_dfs(codomain(neighbor_morphism), D, visited, path)
+        helper_dfs(codomain(neighbor_morphism), D, union(visited, Set([neighbor_morphism])), union(path, [neighbor_morphism]))
+        #helper_dfs(codomain(neighbor_morphism), D, visited, union(path, [neighbor_morphism]))
+      end
+    end
+  end
+
+  helper_dfs(N, M, Set(), [])
+
+  morphisms = Vector{ModuleMap}()
+  for path in all_paths
+    phi = identity_map(N)
+    for h in path
+      phi = phi*h
+    end
+    push!(morphisms, phi)
+  end
+
+  return morphisms
+end
+
+#############################
+# Useful functions
+#############################
+
+@doc Markdown.doc"""
+  register_morphism!(f::ModuleMap)
+Cache the morphism `f` in the corresponding caches of the domain and codomain of `f`.
+"""
+function register_morphism!(f::ModuleMap)
+  push!(domain(f).outgoing_morphisms, f)
+  push!(codomain(f).ingoing_morphisms, f)
+end
+
+#############################
+#TODO move to Hecke
+#  re-evaluate and use or not
+function differential(c::Hecke.ChainComplex, i::Int)
+  return map(c,length(c)-i)
+end
+
+function module_in_complex(c::Hecke.ChainComplex, i::Int)
+  return obj(c,length(c)-i)
+end
+
+getindex(c::Hecke.ChainComplex, i::Int) = module_in_complex(c,i)
+
+function Base.getindex(r::Hecke.SRow, u::UnitRange)
+  R = base_ring(r)
+  s = sparse_row(R)
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function Base.getindex(r::Hecke.SRow, R::AbstractAlgebra.Ring, u::UnitRange)
+  s = sparse_row(R)
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function getindex(a::Hecke.SRow, b::AbstractArray{Int, 1})
+  if length(a.pos) == 0
+    return a
+  end
+  m = minimum(b)
+  b = sparse_row(parent(a.values[1]))
+  for (k,v) = a
+    if k in b
+      push!(b.pos, k-b+1)
+      push!(b.values, v)
+    end
+  end
+  return b
+end
+
+@doc Markdown.doc"""
+  sparse_row(A::MatElem)
+Convert `A` to a sparse row. 
```

Comment:
Always leave an empty line between the list of signatures and the description text. This affects a bunch of places
```suggestion
    sparse_row(A::MatElem)

Convert `A` to a sparse row. 
```